src/java/org/apache/cassandra/cql3/CqlConstraintFunctionExecutor.java

```diff
@@ -0,0 +1,72 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.cassandra.cql3;
+
+
+import java.util.List;
+import java.util.Map;
+
+import org.apache.cassandra.schema.TableMetadata;
+
+/**
+ * Interface to be implemented by functions that are executed as part of CQL constraints.
+ */
+public interface CqlConstraintFunctionExecutor
+{
+    /**
+     * This method returns the function name to be executed.
+     *
+     * @return
+     */
+    String getName();
+
+    /**
+     * Method that provides the execution of the condition. It can either succeed or throw a {@link ConstraintViolationException}.
+     *
+     * @param args
+     * @param relationType
+     * @param term
+     * @param tableMetadata
+     * @param columnValues
+     */
+    void checkConstraint(List<ColumnIdentifier> args, Operator relationType, String term, TableMetadata tableMetadata, Map<String, String> columnValues);
+
+    /**
+     * Method that validates that a condition is valid. This method is called when the CQL constraint is created to determine
+     * if the CQL statement is valid or needs to be rejected as invalid.
+     * @param args
+     * @param relationType
+     * @param term
+     * @param tableMetadata
+     */
+    void validate(List<ColumnIdentifier> args, Operator relationType, String term, TableMetadata tableMetadata);
```

Comment:
Can you consolidate with the interface methods in `ConstraintCondition`? In the implementations, they are tightly coupled.

Reply:
I don't know what you mean, can you elaborate? These two interfaces are defining concepts at two different levels. The ConstraintCondition is a generic constraint umbrella, and the CqlConstraintFunctionExecutor needs different parameters to validate/evaluate.

There is some coupling because a ConstraintCondition that has a function (basically, non scalar) needs a valid CqlConstraintFunctionExecutor as well.